crates/sui-indexer-alt/src/handlers/kv_epoch_ends.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use anyhow::{bail, Context, Result};
+use diesel_async::RunQueryDsl;
+use sui_types::{
+    event::SystemEpochInfoEvent,
+    full_checkpoint_content::CheckpointData,
+    transaction::{TransactionDataAPI, TransactionKind},
+};
+
+use crate::{
+    db,
+    models::epochs::StoredEpochEnd,
+    pipeline::{concurrent::Handler, Processor},
+    schema::kv_epoch_ends,
+};
+
+pub struct KvEpochEnds;
+
+impl Processor for KvEpochEnds {
+    const NAME: &'static str = "kv_epoch_ends";
+
+    type Value = StoredEpochEnd;
+
+    fn process(&self, checkpoint: &Arc<CheckpointData>) -> Result<Vec<Self::Value>> {
+        let CheckpointData {
+            checkpoint_summary,
+            transactions,
+            ..
+        } = checkpoint.as_ref();
+
+        let Some(end_of_epoch) = checkpoint_summary.end_of_epoch_data.as_ref() else {
+            return Ok(vec![]);
+        };
+
+        let Some(transaction) = transactions.iter().find(|tx| {
+            matches!(
+                tx.transaction.intent_message().value.kind(),
+                TransactionKind::ChangeEpoch(_) | TransactionKind::EndOfEpochTransaction(_)
+            )
+        }) else {
+            bail!(
+                "Failed to get end of epoch transaction in checkpoint {} with EndOfEpochData",
+                checkpoint_summary.sequence_number,
+            );
+        };
+
+        if let Some(SystemEpochInfoEvent {
+            total_stake,
+            storage_fund_reinvestment,
+            storage_charge,
+            storage_rebate,
+            storage_fund_balance,
+            stake_subsidy_amount,
+            total_gas_fees,
+            total_stake_rewards_distributed,
+            leftover_storage_fund_inflow,
+            ..
+        }) = transaction
+            .events
+            .iter()
+            .flat_map(|events| &events.data)
+            .find_map(|event| {
+                event
+                    .is_system_epoch_info_event()
```

Comment:
using existence `is_system_epoch_info_event` seems not very robust, what about the `safe_mode` method 
https://github.com/MystenLabs/sui/blob/3a397c8915e933d44654aa6a64f25d1cdf34e93b/crates/sui-types/src/sui_system_state/mod.rs#L175

Reply:
We're not using the existence of `is_system_epoch_info_event` to detect `safe_mode` -- the main purpose of this pipeline is to index this event, and calculating `safe_mode` is a side-effect of that.